clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java

```diff
@@ -52,13 +53,54 @@
 
     private final ConfigDef definition;
 
+    private static final String CONFIG_PROVIDERS_CONFIG = "config.providers";
+
+    /**
+     * Construct a configuration with a ConfigDef, the configuration properties, optional {ConfigProviders}.
+     *
+     * The originals is a name-value pair of config provider configs as well as configuration properties. The
+     * value of the configuration can be a variable as defined below or the actual value. This constructor will
+     * first instantiate the ConfigProviders using the config provider configs, then it will find all the
+     * variables in the values of the originals configurations, attempt to resolve the variables using the named
+     * ConfigProviders, and then parse and validate the configurations.
+     *
+     * ConfigProviders can be passed either as configs in the originals map or in the separate configProviderProps
+     * map. If config providers are passed in the configProviderProps any providers in originals map will be
+     * ignored. If ConfigProvider is not provided, the constructor will skip the variable substitution step and
+     * will simply validate and parse the supplied configuration.
+     *
+     * The config "config.providers" is reserved to list the ConfigProviders, "config.providers.<providerName>.class"
```

Comment:
Again, `ConfigProviders` doesn't make sense here. The `config.providers` property is listing the _names_ of the ConfigProviders, whereas the other `config.providers.<providerName>.class` (for example) specifies the name of the `ConfigProvider` implementation class for the named provider.

IMO, we should use the term `ConfigProvider` to mean one of two things:

1. The name of the `ConfigProvider` interface
2. The instances of the `ConfigProvider` implementations that are instantiated by this class. Always using "instances" in these cases will help disambiguate the use of `ConfigProvider`.

Also, it's probably worthwhile to begin this paragraph with:

> The "{@code config.providers}" configuration property and all configuration properties that begin with the "{@code config.providers.}" prefix are reserved. The  "{@code config.providers}" configuration property specifies the names of the config providers, and properties that begin with the  "{@code config.providers.<providerName>.}" prefix correspond to the properties for that named provider. For example, the "{@code config.providers.<providerName>.class}" property specifies the name of the {@link ConfigProvider} implementation class that should be used for the provider.